app/public/generation.py

```diff
@@ -43,19 +44,38 @@ async def user_identifier(request):
     return request.cookies.get("uuid")
 
 @router.post("/generate/{challenge_id}", dependencies=[Depends(RateLimiter(times=public_settings.GEN_REQUESTS_PER_MINUTE, minutes=1, identifier=global_identifier))])
-async def generate(challenge_id: int, request: GenerateRequest) -> GenerateResponse:
+async def generate(challenge_id: int, request: GenerateRequest, uuid: Annotated[str | None, Cookie()] = None) -> GenerateResponse:
+    uuid = get_user_from_cookie(uuid)
+    if uuid is None:
+        return GenerateResponse(error="Use the normal index!")
+
     log.info(f"Generating with prompt.")
     with SessionLocal() as session:
-        # Check this isn't a duplicate
-        generation = session.query(Generation).filter(Generation.challenge_id == challenge_id).filter(Generation.prompt == request.prompt).first()
-        if generation:
-            log.info("Duplicate generation found")
-            return GenerateResponse(error="Duplicate prompt found, try something else.")
         # Check that the model is active
         model = session.query(Model).filter(Model.name == request.model).first()
         if model is not None and not model.active:
             log.error(f"Model {model.name} is not active!")
             return GenerateResponse(error="Model not active!")
+
+        # Check this isn't a duplicate
+        generation = session.query(Generation).filter(Generation.challenge_id == challenge_id).filter(Generation.prompt == request.prompt, Generation.model_id == model.id).first()
+        if generation:
```

Comment:
Check the tab indentation because it could be wrong. Im breaking apart this if statement block so its very clear which block should be hit in which circumstance

```suggestion
        if generation and Generation.usr_uuid == uuid:
            log.info("Duplicate generation found for the same model for this user.  Returning cached response.")
       
        if generation and Generation.usr_uuid ! uuid: # See comment above about bug
                # A different user has already generated this prompt for this model, so we can reuse the generation.
                new_generation = Generation(
                    challenge_id=challenge_id,
                    model_id=model.id,
                    prompt=request.prompt,
                    generation=generation.generation,
                    usr_uuid=uuid,
                )
                session.add(new_generation)
                session.commit()
                return GenerateResponse(generation=new_generation.generation, id=new_generation.id)
       
```